app/src/main/java/org/wikipedia/page/PageActivity.kt

```diff
@@ -794,6 +820,110 @@ class PageActivity : BaseActivity(), PageFragment.Callback, LinkPreviewDialog.Lo
+
+                        MaterialAlertDialogBuilder(this@PageActivity)
+                            .setTitle(R.string.suggested_reading_list_dialog_title)
+                            .setMessage(R.string.suggested_reading_list_dialog_body)
+                            .setPositiveButton(R.string.suggested_reading_list_dialog_positive) { _, _ ->
+                                RabbitHolesEvent.submit("enter_click", "reading_list_prompt")
+                                requestSuggestedReadingListLauncher.launch(ReadingListActivity.newIntent(this@PageActivity, true, suggestedList = true))
+                            }
+                            .setNegativeButton(R.string.suggested_reading_list_dialog_negative) { _, _ ->
+                                RabbitHolesEvent.submit("ignore_click", "reading_list_prompt")
+                                FeedbackUtil.showMessage(this@PageActivity, R.string.suggested_reading_list_later_snackbar)
+                            }
+                            .show()
+                    }
+                }
+            }
+        }
+        maybeShowRabbitHolesSurvey()
+    }
+
+    private fun applySuggestedSearchTerm(term: String) {
+        suggestedSearchTerm = term
+        binding.pageToolbarButtonSearch.text = term
+    }
+
+    private fun applySuggestedReadingList(basedOnTitle1: HistoryEntry, basedOnTitle2: HistoryEntry, pages: List<MwQueryPage>) {
+        val listItems = pages.map {
+            JsonUtil.json.encodeToJsonElement(
+                ReadingListsShareHelper.ExportedReadingListPage(
+                    basedOnTitle1.title.wikiSite.languageCode,
+                    it.displayTitle(basedOnTitle1.title.wikiSite.languageCode),
+                    it.ns,
+                    it.description,
+                    it.thumbUrl()
+                )
+            )
+        }
+        val readingList = ReadingListsShareHelper.ExportedReadingList(
+            list = mapOf(basedOnTitle1.title.wikiSite.languageCode to listItems),
+            name = getString(R.string.suggested_reading_list_title),
+            description = getString(R.string.suggested_reading_list_description_multi,
+                StringUtil.fromHtml(basedOnTitle1.title.displayText).toString(),
+                StringUtil.fromHtml(basedOnTitle2.title.displayText).toString())
+        )
+        Prefs.importReadingListsDialogShown = false
+        Prefs.suggestedReadingListsData = JsonUtil.encodeToString(readingList)
+    }
+
+    private fun maybeShowRabbitHolesSurvey() {
+        if (Prefs.suggestedContentSurveyShown) {
+            return
+        }
+        lifecycleScope.launch(CoroutineExceptionHandler { _, t ->
+            L.e(t)
+        }) {
+            delay(TimeUnit.SECONDS.toMillis(if (ReleaseUtil.isDevRelease) 1L else 10L))
+            pageFragment.historyEntry?.let {
+                if (!Prefs.suggestedContentSurveyShown && it.source == HistoryEntry.SOURCE_RABBIT_HOLE_SEARCH) {
```

Comment:
Can we move this to the line `905`? This seems duplicated.

Reply:
Sure.